### PR TITLE
Int 410 fix logout callback

### PR DESCRIPTION
**Pull request by saraOSI**
The blur on the profile avatar is preventing the logoutCallback from being called. Make sure that the blur event doesn't do that. More info in the messages on a commit-by-commit basis.

This is the least-bad solution I could come up with. Other suggestions on the interwebs included using`mousedown` instead of `onclick` for the callback. But that seems like a bad idea, imo. Other, even less less bad solutions, are more than welcome.

### CODE REVIEW

**Comment by SashaKyz** [comment]
Can one of the admins verify this patch?

File: src/test/ProfileAvatar_test.js [hunk 1]
```diff
@@ -50,8 +50,8 @@ describe('ProfileAvatar', () => {
       );
     });
 
-    it('blurring the profile avatar hides visible logout button', () => {
-      profileAvatar.setState({ isHidden: false });
+    it('blurring the avatar when the menu is not moused over hides the logout button', () => {
```
Comment (edgibbs) [inline src/test/ProfileAvatar_test.js]: 'blurs the avatar'

File: src/test/ProfileAvatar_test.js [hunk 2]
```diff
@@ -50,14 +50,77 @@ describe('ProfileAvatar', () => {
       );
     });
 
-    it('click on logout invokes logoutCallback', () => {
+    it('blurring the avatar when the menu is not moused over hides the logout button', () => {
```
Comment (edgibbs) [inline src/test/ProfileAvatar_test.js]: Small thing, but I'd suggest nesting a describe block here similar to using context's in rspec,

```javascript
describe('when the menu is not moused over', () => {
    it('blurs the avatar', () => {
  });
})l

File: src/test/ProfileAvatar_test.js [hunk 3]
```diff
@@ -50,14 +50,77 @@ describe('ProfileAvatar', () => {
...
+        .simulate('blur', { preventDefault: () => {} });
+      expect(profileAvatar.findWhere(n => n.text() === 'Logout').exists()).toBe(
+        true
+      );
+    });
+
+    it('clicking on logout invokes logoutCallback', () => {
       profileAvatar.setProps({ logoutCallback: spyCallback });
       profileAvatar.setState({ isHidden: false });
+      // clicking the logout link also mouses over the link and blurs the avatar
```
Comment (edgibbs) [inline src/test/ProfileAvatar_test.js]: remove comment

File: src/test/ProfileAvatar_test.js [hunk 4]
```diff
@@ -50,14 +50,77 @@ describe('ProfileAvatar', () => {
...
+      profileAvatar
+        .find('a')
+        .first()
+        .simulate('blur', { preventDefault: () => {} });
+      expect(profileAvatar.findWhere(n => n.text() === 'Logout').exists()).toBe(
+        false
+      );
+    });
+
+    it('blurring the avatar when the menu is moused over does not hide the logout button', () => {
```
Comment (edgibbs) [inline src/test/ProfileAvatar_test.js]: Similiar comment here with nesting a describe block.

File: src/test/ProfileAvatar_test.js [hunk 5]
```diff
@@ -50,14 +50,77 @@ describe('ProfileAvatar', () => {
...
+      profileAvatar
+        .find('a')
+        .first()
+        .simulate('blur', { preventDefault: () => {} });
+      expect(profileAvatar.findWhere(n => n.text() === 'Logout').exists()).toBe(
+        true
+      );
+    });
+
+    it('clicking on logout invokes logoutCallback', () => {
```
Comment (edgibbs) [inline src/test/ProfileAvatar_test.js]: another describe('when clicking on logout') it('invokes logoutCallback')

File: src/test/ProfileAvatar_test.js [hunk 6]
```diff
@@ -50,14 +50,77 @@ describe('ProfileAvatar', () => {
...
+        .simulate('blur', { preventDefault: () => {} });
+      profileAvatar
+        .find('a')
+        .at(1)
+        .simulate('click', { preventDefault: () => {} });
       expect(spyCallback).toHaveBeenCalled();
+      expect(profileAvatar.html()).not.toContain('logout');
+    });
+
+    it('clicking on logout closes the dropdown menu and sets dropdownFocused to false', () => {
```
Comment (edgibbs) [inline src/test/ProfileAvatar_test.js]: Same describe for context in the next 3.

File: src/test/ProfileAvatar_test.js (hunk 2 above)
Comment (saraOSI) [reply src/test/ProfileAvatar_test.js]: I thought about doing this and didn't for some strange reason. I can make that fix.

File: src/test/ProfileAvatar_test.js (hunk 6 above)
Comment (saraOSI) [reply src/test/ProfileAvatar_test.js]: I'm not sure all of these test belong in the same context, but I moved them into the contexts I thought appropriate.

File: src/ProfileAvatar.js [hunk 7]
```diff
@@ -30,9 +34,15 @@ class ProfileAvatar extends React.Component {
...
-            onClick={event =>
-              this.props.logoutCallback(this.props.profileId, event)
-            }
+            onMouseOver={() => this.setState({ dropdownFocused: true })}
+            onFocus={() => this.setState({ dropdownFocused: true })}
+            onMouseOut={() => this.setState({ dropdownFocused: false })}
+            onBlur={() => this.setState({ dropdownFocused: false })}
+            onClick={event => {
+              event.preventDefault();
+              this.props.logoutCallback(this.props.profileId, event);
```
Comment (ratneshraval) [inline src/ProfileAvatar.js]: So, it seems like we don't really need event here or in the logoutCallback().
How about,
```
onClick={ 
this.props.logoutCallback(this.props.profileId)
this.setState({ dropdownFocused: false, isHidden: true });
}
```

Comment (saraOSI) [reply src/ProfileAvatar.js]: Unfortunately, this component is using `a` tags, which means if we don't `preventDefault()` on the `event`, the URL for the current page changes. Adding an anchor tag isn't the end of the world, but I would prefer for links to not change the URL unless necessary.

Comment (ratneshraval) [reply src/ProfileAvatar.js]: even if it was `<a href="#" ` ?

**Comment by edgibbs** [comment]
retest this please

File: src/ProfileAvatar.js (hunk 7 above)
Comment (saraOSI) [reply src/ProfileAvatar.js]: I believe so. An empty anchor will also scroll the page to the top, in the instance that we ever make the global header stick as a user scrolls.

**Pull request merged**